src/core/data.jl

```diff
@@ -2526,16 +2526,34 @@ function calc_connected_components(data::Dict{String,<:Any}; edges=["branch", "d
         end
     end
 
-    component_lookup = Dict(i => Set{Int}([i]) for i in active_bus_ids)
-    touched = Set{Int}()
-
-    for i in active_bus_ids
-        if !(i in touched)
-            _cc_dfs(i, neighbors, component_lookup, touched)
+    sorted_bus_ids = sort(collect(active_bus_ids))
+    i0 = sorted_bus_ids[1] - 1  # this is to track un-visited buses
+    # The two dictionaries below are used to track the connected components
+    # `component_lookup` maps each bus to the ID of the connected component it belongs to, which is the smallest bus ID in said components
+    # `components` maps the connected component ID to a `Set{Int}` of all bus IDs in that component
+    component_lookup = Dict(i => i0 for i in active_bus_ids)
+    components = Dict{Int,Set{Int}}()
+
+    # ⚠️ it is important to iterate over _sorted_ bus IDs to ensure that components are labeled correctly  
+    for i in sorted_bus_ids
+        component_lookup[i] != i0 && continue  # bus already flagged; skip
+
+        # We have a new connected component!
+        component_lookup[i] = i
+        components[i] = Set([i])
+        V = [i]  # list of all bus IDs in this connected component
+        while length(V) > 0
+            j = pop!(V)
+            for k in neighbors[j]
+                component_lookup[k] == i0 || continue  # already visited
```

Comment:
Can you remind me how this Julia `||` syntax works? I am _guessing_ that it means, if the lhs is false then execute the rhs?

Is it also the case that `component_lookup[k] == i` if it does not equal `i0`, if not, then we have an issue where a previous labeling did not explore the graph correctly, right?

Reply:
The syntax is equivalent to 
```julia
if component_lookup[k] == i0
    # do nothing
else
    continue
end
```

Reply:
> Is it also the case that component_lookup[k] == i if it does not equal i0, if not, then we have an issue where a previous labeling did not explore the graph correctly, right?

If `component_lookup[k] != i0`, then it should be equal to the index of the bus with smallest index in that connected component. I believe it should then be equal to `i`, yes, but only 95% confident.

Reply:
If it not the case 100% of the time then I worry the implementation has a correctness issue. 

How about we add an assertion after this check that `component_lookup[k] == i`, this will help catch the issue down the road if we run into it.